google/resource_dataproc_cluster.go

```diff
@@ -0,0 +1,936 @@
+					sort.Strings(cluster.Config.GceClusterConfig.ServiceAccountScopes)
+				}
+			}
+
+			if cfg, ok := configOptions(d, "cluster_config.0.software_config"); ok {
+				cluster.Config.SoftwareConfig = &dataproc.SoftwareConfig{}
+
+				if v, ok := cfg["override_properties"]; ok {
+					m := make(map[string]string)
+					for k, val := range v.(map[string]interface{}) {
+						m[k] = val.(string)
+					}
+					cluster.Config.SoftwareConfig.Properties = m
+				}
+				if v, ok := cfg["image_version"]; ok {
+					cluster.Config.SoftwareConfig.ImageVersion = v.(string)
+				}
+			}
+
+			if v, ok := d.GetOk("cluster_config.0.initialization_action"); ok {
+				actionList := v.([]interface{})
+
+				actions := []*dataproc.NodeInitializationAction{}
+				for _, v1 := range actionList {
+					actionItem := v1.(map[string]interface{})
+					action := &dataproc.NodeInitializationAction{
+						ExecutableFile: actionItem["script"].(string),
+					}
+					if x, ok := actionItem["timeout_sec"]; ok {
+						action.ExecutionTimeout = strconv.Itoa(x.(int)) + "s"
+					}
+
+					actions = append(actions, action)
+				}
+				cluster.Config.InitializationActions = actions
+			}
+
+			if cfg, ok := configOptions(d, "cluster_config.0.master_config"); ok {
+				log.Println("[INFO] got master_config")
+				cluster.Config.MasterConfig = instanceGroupConfigCreate(cfg)
+			}
+
+			if cfg, ok := configOptions(d, "cluster_config.0.worker_config"); ok {
+				log.Println("[INFO] got worker config")
+				cluster.Config.WorkerConfig = instanceGroupConfigCreate(cfg)
+			}
+
+			if cfg, ok := configOptions(d, "cluster_config.0.preemptible_worker_config"); ok {
+				log.Println("[INFO] got preemtible worker config")
+				cluster.Config.SecondaryWorkerConfig = preemptibleInstanceGroupConfigCreate(cfg)
+				if cluster.Config.SecondaryWorkerConfig.NumInstances > 0 {
+					cluster.Config.SecondaryWorkerConfig.IsPreemptible = true
+				}
+			}
+		}
+	}
+
+	// Checking here caters for the case where the user does not specify cluster_config
+	// at all, as well where it is simply missing from the gce_cluster_config
+	if region == "global" && !zok {
+		return errors.New("zone is mandatory when region is set to 'global'")
+	}
+
+	// Create the cluster
+	op, err := config.clientDataproc.Projects.Regions.Clusters.Create(
+		project, region, cluster).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(clusterName)
+
+	// Wait until it's created
+	timeoutInMinutes := int(d.Timeout(schema.TimeoutCreate).Minutes())
+	waitErr := dataprocClusterOperationWait(config, op, "creating Dataproc cluster", timeoutInMinutes, 3)
+	if waitErr != nil {
+		// The resource didn't actually create
+		d.SetId("")
+		return waitErr
+	}
+
+	log.Printf("[INFO] Dataproc cluster %s has been created", clusterName)
+	return resourceDataprocClusterRead(d, meta)
+
+}
+
+func preemptibleInstanceGroupConfigCreate(cfg map[string]interface{}) *dataproc.InstanceGroupConfig {
```

Comment:
We usually call functions like this expand functions, so this would be called `expandPreemptibleInstanceGroupConfig`. Mind renaming these?